File: 0043-Multiply Strings/0043-Multiply Strings.cpp
class Solution {
public:
    string multiply(string num1, string num2) {
        int m = num1.size();
        int n = num2.size();
        vector<int> product(m + n);
        for (int i = m - 1; i >= 0; --i) {
            for (int j = n - 1; j >= 0; --j) {
                int p1 = i + j;
                int p2 = i + j + 1;
                int sum = (num1[i] - '0') * (num2[j] - '0') + product[p2];
                product[p1] += sum / 10;
                product[p2] = sum % 10;
            }
        }
        
        int i = 0;
        while (i < product.size() - 1 && product[i] == 0) {
            ++i;
        }

        string result;
        while (i < product.size()) {
            result += '0' + product[i++];
        }
        return result;
    }
};

File: 0001-0100/0043-Multiply Strings/0043-Multiply Strings.py
class Solution:
    def multiply(self, num1: str, num2: str) -> str:
        product = [0] * (len(num1) + len(num2))
        for i in range(len(num1) - 1, -1, -1):
            for j in range(len(num2) - 1, -1, -1):
                p1 = i + j
                p2 = i + j + 1
                sum = int(num1[i]) * int(num2[j]) + product[p2]
                product[p1] += sum // 10
                product[p2] = sum % 10
                
        return ''.join(map(str, product)).lstrip('0') or '0'
